### PR TITLE
[MM-44347] Add popover for user groups

### PR DESCRIPTION
**Pull request by cyrilzhang-mm**
#### Summary

Adds a user group popover component that is activated by group at-mentions. I also tested and verified it with VoiceOver on Google Chrome, macOS. 

#### Ticket Link

[Jira ticket](https://mattermost.atlassian.net/browse/MM-44347)

#### Related Pull Requests

- Has server changes: https://github.com/mattermost/mattermost-server/pull/21523

The above server PR changes the order that group members are returned in, but it's not necessary for this PR to work.

#### Screenshots

Small group:
Screen Shot 2022-10-26 at 4 20 17 PM

Hover state:
Screen Shot 2022-10-26 at 4 22 08 PM

Onyx theme:
Screen Shot 2022-10-26 at 5 46 30 PM

Large group:
Screen Shot 2022-10-26 at 4 20 35 PM

Large group, no search results:
Screen Shot 2022-10-26 at 4 20 52 PM

Large group with search result:
Screen Shot 2022-10-26 at 4 20 46 PM

No edit permission (on Firefox so the scrollbar looks different):
Screen Shot 2022-10-26 at 5 49 42 PM

#### Release Note

```release-note
Added a popover containing group information to group at-mentions.
```

### CODE REVIEW

File: components/user_group_popover/user_group_popover.tsx [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    padding: 16px 20px;
+`;
+
+const Title = styled.span`
+    flex: 1 1 auto;
+`;
+
+const EditButton = styled.button`
+    width: 24px;
+    height: 24px;
```
Comment (BenCookie95) [inline components/user_group_popover/user_group_popover.tsx:195]: I think the hover effect is a little small and needs to be 28px, that would match it with the figma

File: components/user_group_popover/user_group_popover.tsx [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                            }}
+                            tagName={NoShrink}
+                        />
+                    </Subtitle>
+                    {group.member_count > 10 ? (
+                        <SearchBar
+                            type='text'
+                            className='user-group-popover_search-bar'
+                            placeholder={Utils.localizeMessage('user_group_popover.searchGroupMembers', 'Search members')}
+                            inputPrefix={<i className={'icon icon-magnify'}/>}
```
Comment (BenCookie95) [inline components/user_group_popover/user_group_popover.tsx:164]: The padding-left on the prefix should be 11px

**Review by BenCookie95 — CHANGES_REQUESTED** [review]
This is amazing, can't wait to use it once it's merged. 2 very minor style comments

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Nice work @cyrilzhang-mm! This is great. A few minor things from me, but it's really close:

1. The group display name in the popover should be using `Metropolis` as the font
2. For the • character that divides the mention name and members count, can we reduce the space on either side of the • to 6 instead of 8?
3. Can you make the size of the pencil icon button 28x28 instead of 24x24? The icon itself is correct, but the button container should be 28x28. You may need to adjust the alignment of the icon within button as well as the position within the header after you make that change. Changing this may also affect the space between group title and mention name below so that may need some adjustment too.
4. Can you make the border that divides the header and the user list 8% opacity instead of 16%?
5. For the search input, it looks like you'll need to use the smaller version of the search input. The design has an input with 32px height, but this looks to be 40px in height. In addition, the font-size and padding within the search input seem not quite right. Also, you can add a margin of `margin: 0 -6px` to the input container so that the field spans a bit wider as designed

image

**Comment by cyrilzhang-mm** [comment]
Updated to address design feedback. Screenshot:
Screen Shot 2022-10-31 at 11 40 34 AM

**Review by matthewbirtch — APPROVED** [review]
UI updates look good @cyrilzhang-mm, thanks for resolving those!

**Review by BenCookie95 — APPROVED** [review]

**Review by mkraft — APPROVED** [review]

File: reducers/views/search.ts [hunk 3]
```diff
@@ -23,6 +23,19 @@ function modalSearch(state = '', action: GenericAction) {
     }
 }
 
+function popoverSearch(state = '', action: GenericAction) {
+    switch (action.type) {
+    case SearchTypes.SET_POPOVER_SEARCH: {
+        return action.data.trim();
+    }
+
+    case UserTypes.LOGOUT_SUCCESS:
```
[the remark this answers is not in the document]
Comment (cyrilzhang-mm) [reply reducers/views/search.ts:32]: I think I just copied it from the `modalSearch` above, but I think you're right, it's not necessary.

File: components/user_group_popover/group_member_list/group_member_list.tsx [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+            if (variableSizeListRef.current) {
+                variableSizeListRef.current.resetAfterIndex(0);
+            }
+        }
+        setHasMounted(true);
+    }, [users.length, hasMounted]);
+
+    const loadNextPage = async () => {
+        setIsNextPageLoading(true);
+        const res = await actions.getUsersInGroup(group.id, nextPage, USERS_PER_PAGE);
```
[the remark this answers is not in the document]
Comment (cyrilzhang-mm) [reply components/user_group_popover/group_member_list/group_member_list.tsx:115]: Yes, that's true. Alternatively if we have no design for the failure case I can make it stay as the loading indicator?

[the remark this answers is not in the document]
Comment (cyrilzhang-mm) [reply components/user_group_popover/group_member_list/group_member_list.tsx:115]: @matthewbirtch and I have discussed this and decided on a design, which has been added to Figma.

**Comment by cyrilzhang-mm** [comment]
Marking PR as draft while I work on the design updates.

File: components/user_group_popover/group_member_list/group_member_list.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthewbirtch) [reply components/user_group_popover/group_member_list/group_member_list.tsx:115]: @mkraft FYI I've been discussing this with @cyrilzhang-mm and these should be addressed in the updates

**Comment by cyrilzhang-mm** [comment]
Updated with the new designs. Sorry about the number of commits, I had some merge issues. 

Screen Shot 2022-11-09 at 2 35 38 PM

Screen Shot 2022-11-09 at 2 35 54 PM

Screen Shot 2022-11-09 at 2 36 53 PM

Screen Shot 2022-11-09 at 2 36 34 PM

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Nice work @cyrilzhang-mm. It's looking great with the changes you made.

One small thing:
1. Can we add the clear `x` icon within the search field on the right side so that users can clear the search field with one click (like they can in global search)

image

File: components/user_group_popover/user_group_popover.tsx [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+`;
+
+const CloseButton = styled.button`
+    width: 28px;
+    height: 28px;
+    flex: 0 0 auto;
+    margin-left: 4px;
+    display: flex;
+    justify-content: center;
+    align-items: center;
```
Comment (matthewbirtch) [inline components/user_group_popover/user_group_popover.tsx:230]: Can we bump the close button 4 pixels to the right? One way to do that:
```
position: relative;
right: -4px;
```

**Comment by cyrilzhang-mm** [comment]
@matthewbirtch I updated the close button and search bar. I ended up using a different input component for the search bar, so I had to re-style it again. I tried to match it to the previous one as well as the text input Figma document as much as possible but please let me know if I missed anything!

File: components/user_group_popover/user_group_popover.tsx [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        color: var(--center-channel-color);
+        flex: 1;
+    }
+
+    input.a11y--focused {
+        box-shadow: none;
+    }
+
+    svg {
+        width: 14px;
```
Comment (matthewbirtch) [inline components/user_group_popover/user_group_popover.tsx:310]: adjust width to 18px

File: components/user_group_popover/user_group_popover.tsx [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        height: 100%;
+        margin: 0 11px;
```
Comment (matthewbirtch) [inline components/user_group_popover/user_group_popover.tsx:312]: change margin to 
```
margin: 0 6px;
```

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Just a few minor tweaks and I think we're good!

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Sorry, just tested a dark theme and it looks like there is an issue in the search input on dark themes: 

image

I think you just may need to put the `center-channel-bg` color as a background color on the input

**Comment by cyrilzhang-mm** [comment]
Thanks for catching that @matthewbirtch!

**Review by matthewbirtch — APPROVED** [review]

**Review by mkraft — APPROVED** [review]

**Review by BenCookie95 — APPROVED** [review]
LGTM!

**Comment by furqanmlk** [comment]
Hi @BenCookie95  can you please conflict

**Comment by cyrilzhang-mm** [comment]
@furqanmlk Fixed the conflicts, sorry about that!

**Review by furqanmlk — COMMENTED** [review]
I see the following issues

1. Name is not sorted, when first name and last name is selected in profile settings.
image

2. Popover is hiding profile picture and merging into LHS 
**Figma**
image
**PR**
image

**Comment by cyrilzhang-mm** [comment]
@furqanmlk Pushed a PR that will sort users by display name. It should fix the first issue.

I believe the second one isn't an issue. It behaves the same way as the existing profile popovers.

**Review by furqanmlk — COMMENTED** [review]
Hi @cyrilzhang-mm , I noticed few ui issues

1.In  Search loading state, spinner icon is different
 **PR**

image

**Figma**
image

2. In no search result, search icon is different
**PR**
image

**Figma**
image

3. Can u look at  the gif to confirm user group popover ui
the first group id popover is hiding avater
**PR**

https://user-images.githubusercontent.com/6909890/203131734-21d8ed77-90c7-4060-96df-86ee6512f260.mp4

**Comment by cyrilzhang-mm** [comment]
Hi @furqanmlk the gif looks good to me, the popovers are centered with the mention link. It's a limitation of the popover library we use and it matches the behaviour of the profile popover. 

The icons are okay too. They are the same icons as what's already used in the app which @matthewbirtch approved. Thanks :)

**Pull request merged**